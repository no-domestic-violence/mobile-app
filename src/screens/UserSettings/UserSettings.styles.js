import { StyleSheet } from 'react-native';
import { Colors } from 'styles';

export const styles = StyleSheet.create({
  button: {
    backgroundColor: Colors.primary,
    paddingVertical: 5,
    paddingHorizontal: 15,
    borderWidth: 0,
    borderRadius: 14,
    color: '#000',
    fontWeight: '500',
    fontSize: 20,
    overflow: 'hidden',
    textAlign: 'center',
    width: '40%',
    marginLeft: 10,
  },
  userSettingsContainer: {
    alignItems: 'flex-start',
  },
  userSettingsActions: {
    fontSize: 20,
    fontWeight: '500',
    lineHeight: 40,
  },
  userSettingsActionsView: {
    marginVertical: 50,
    marginLeft: 10,
  },
});
