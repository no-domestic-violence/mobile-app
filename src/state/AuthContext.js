import AsyncStorage from '@react-native-async-storage/async-storage';
import createAppContext from './CreateAppContext';
import appApiClient from '../api/appApiClient';

const authReducer = (state, action) => {
  switch (action.type) {
    case 'SIGNUP_ERROR':
    case 'LOGIN_ERROR':
      return {
        ...state,
        errorMessage: action.payload,
      };
    case 'SIGNUP_SUCCESS':
    case 'LOGIN_SUCCESS':
      return {
        ...state,
        token: action.payload.token,
        username: action.payload.user.username,
        errorMessage: '',
      };
    case 'CHANGE_PASSWORD_SUCCESS':
      return {
        ...state,
        successMessage: action.payload,
      };
    case 'CHANGE_PASSWORD_ERROR':
      return {
        ...state,
        errorMessage: action.payload,
      };
    case 'AUTH_SUCCESS':
      return {
        ...state,
        token: action.payload.token,
        username: action.payload.username,
      };
    case 'LOGOUT':
      return {
        token: '',
        username: '',
      };
    case 'DELETE_ACCOUNT':
      return {
        token: '',
        username: '',
        successMessage: action.payload.data,
      };
    case 'REMOVE_ERRORS':
      return {
        ...state,
        errorMessage: '',
      };
    case 'REMOVE_MESSAGES':
      return {
        ...state,
        successMessage: '',
      };
    default:
      return state;
  }
};

const signup = (dispatch) => async ({ email, password, username }) => {
  try {
    const response = await appApiClient.post('/signup', {
      email,
      password,
      username,
    });

    await AsyncStorage.setItem('token', response.data.token);
    await AsyncStorage.setItem('username', response.data.user.username);

    dispatch({ type: 'SIGNUP_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'SIGNUP_ERROR',
      payload: "Something went wrong'( Try again",
    });
  }
};

const login = (dispatch) => async ({ email, password }) => {
  try {
    const response = await appApiClient.post('/login', { email, password });
    await AsyncStorage.setItem('token', response.data.token);
    await AsyncStorage.setItem('username', response.data.user.username);
    dispatch({ type: 'LOGIN_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'LOGIN_ERROR',
      payload: 'Are you sure about password and email?',
    });
  }
};

const authentication = (dispatch) => async () => {
  const token = await AsyncStorage.getItem('token');
  const username = await AsyncStorage.getItem('username');
  const user = { token, username };
  if (token) {
    dispatch({ type: 'AUTH_SUCCESS', payload: user });
  }
};
const removeErrors = (dispatch) => () => {
  dispatch({ type: 'REMOVE_ERRORS' });
};

const removeMessages = (dispatch) => () => {
  dispatch({ type: 'REMOVE_MESSAGES' });
};

const changePassword = (dispatch) => async ({
  email,
  password,
  oldPassword,
}) => {
  try {
    const response = await appApiClient.post('/changePassword', {
      email,
      oldPassword,
      password,
    });
    removeErrors();
    dispatch({ type: 'CHANGE_PASSWORD_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'CHANGE_PASSWORD_ERROR',
      payload: error.response.data.message,
    });
  }
};

const signout = (dispatch) => async () => {
  await AsyncStorage.removeItem('token');
  dispatch({ type: 'LOGOUT' });
};

const deleteAccount = (dispatch) => async ({ username }) => {
  try {
    // TODO: fix sending params so it is more readable
    const response = await appApiClient.delete('/deleteUser', {
      params: { username },
    });
    await AsyncStorage.removeItem('token');
    await AsyncStorage.removeItem('username');
    dispatch({ type: 'DELETE_ACCOUNT', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'CHANGE_PASSWORD_ERROR',
      payload: error.response.data.message,
    });
  }
};

export const { Provider, Context } = createAppContext(
  authReducer,
  {
    signup,
    login,
    signout,
    removeErrors,
    removeMessages,
    authentication,
    changePassword,
    deleteAccount,
  },
  { isLoggedIn: false, errorMessage: '' }
);
