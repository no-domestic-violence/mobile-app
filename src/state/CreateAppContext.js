import React, { useReducer } from 'react';

// boilerplate for creating Context and Provider
// does not to be changed
export default (reducer, actions, defaultValue) => {
  const Context = React.createContext();

  const Provider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, defaultValue);

    //actions is object for ex {updateCurrentLocation: (dispatch)=> (location) => { dispatch({ type: 'UPDATE_LOCATION', payload: location });}
    const boundActions = {};
    //we need to loop through each action in object and for every key(which is function) we call it with dispatch argument and it gives back function. this function we pass down to all children like ...boundActions
    for (const key in actions) {
      boundActions[key] = actions[key](dispatch);
    }
    return (
      <Context.Provider value={{ state, ...boundActions }}>
        {children}
      </Context.Provider>
    );
  };
  return { Context, Provider };
};
