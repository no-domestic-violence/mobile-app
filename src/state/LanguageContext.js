import React, { createContext, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import i18n from 'services/i18n';
import PropTypes from 'prop-types';
import { DEFAULT_LANGUAGE } from '../services/i18n/language-detector';

const APP_LANGUAGE = 'lang';

const availableLanguages = [
  { lang: 'en', name: 'English' },
  { lang: 'de', name: 'German' },
];

export const LanguageContext = createContext({
  availableLanguages,
  appLanguage: DEFAULT_LANGUAGE,
  setAppLanguage: () => {},
  initializeAppLanguage: () => {},
});

export const LanguageProvider = ({ children }) => {
  const [appLanguage, setAppLanguage] = useState(DEFAULT_LANGUAGE);

  const setLanguage = async (language) => {
    i18n.changeLanguage(language);
    setAppLanguage(language);
    AsyncStorage.setItem(APP_LANGUAGE, language);
  };

  const initializeAppLanguage = async () => {
    const currentLanguage = await AsyncStorage.getItem(APP_LANGUAGE);
    if (!currentLanguage) {
      setLanguage(DEFAULT_LANGUAGE);
    } else {
      setLanguage(currentLanguage);
    }
  };

  return (
    <LanguageContext.Provider
      value={{
        availableLanguages,
        setAppLanguage: setLanguage,
        appLanguage,
        initializeAppLanguage,
      }}>
      {children}
    </LanguageContext.Provider>
  );
};
LanguageProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
