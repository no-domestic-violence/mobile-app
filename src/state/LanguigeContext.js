//TODO: it did not work, to figure out if we need or we can just you provider for everything


// import React, { createContext } from 'react';
// import { I18nextProvider } from 'react-i18next';
// import i18next from 'i18next';

// const LanguigeContext = createContext();

// function LanguigeProvider(props) {

//   return (
//     <LanguigeContext.Provider>
//       <I18nextProvider i18n={i18next}>{props.children}</I18nextProvider>
//    </LanguigeContext.Provider>
//   );
// }

// export {LanguigeProvider, LanguigeContext};

