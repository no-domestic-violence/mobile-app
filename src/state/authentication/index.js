import createAppContext from 'helpers/createAppContext';
import {
  signup,
  login,
  signout,
  removeErrors,
  removeMessages,
  authentication,
  changePassword,
  deleteAccount,
  setAlreadyLaunchedValue,
} from './authentication.actions';
import authReducer from './authentication.reducer';

export const { Provider, Context } = createAppContext(
  authReducer,
  {
    signup,
    login,
    signout,
    removeErrors,
    removeMessages,
    authentication,
    changePassword,
    deleteAccount,
    setAlreadyLaunchedValue,
  },
  { isLoggedIn: false, errorMessage: '', isFirstLaunch: null }
);
