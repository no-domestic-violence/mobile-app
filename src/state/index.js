import React, {createContext, useState, useMemo} from 'react';

const AppContext = createContext();

function ContextProvider(props) {
  //just an example of context usage
  const [userToken, setUserToken] = useState('');
  const authContext = useMemo(() => {
    return {
      logIn: () => {
        setUserToken('tokenishere');
      },
      signUp: () => {
        setUserToken('tokenishere');
      },
      signOut: () => {
        setUserToken('');
      },
    };
  });

  return (
    <AppContext.Provider value={{authContext, userToken}}>
      {props.children}
    </AppContext.Provider>
  );
}

export {ContextProvider, AppContext};
