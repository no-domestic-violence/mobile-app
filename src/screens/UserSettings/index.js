import React, { useContext } from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { Context as AuthContext } from 'state/AuthContext';
import { Colors } from 'styles';
import { StyledView } from 'styles/shared/StyledView';
import UserInfo from '_components/user-settings/UserInfo';
import { useTranslation } from 'react-i18next';

export default function UserSettings({ navigation }) {
  const { t } = useTranslation();
  const { state, signout } = useContext(AuthContext);
  const { username } = state;
  const handleSignOut = () => {
    signout();
  };
  return (
    <StyledView style={styles.userSettingsContainer}>
      <UserInfo username={username} />
      <View style={styles.userSettingsActionsView}>
        <TouchableOpacity onPress={() => navigation.navigate('How to use')}>
          <Text style={styles.userSettingsActions}>
            {t('common.how-to-use')}
          </Text>
        </TouchableOpacity>
        {state.token ? (
          <>
            <TouchableOpacity
              onPress={() => navigation.navigate('Change Password')}>
              <Text style={styles.userSettingsActions}>
                {t('common.change-password')}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => navigation.navigate('Change Language')}>
              <Text style={styles.userSettingsActions}>
                {t('common.change-language')}
              </Text>
            </TouchableOpacity>
            <TouchableOpacity
              onPress={() => navigation.navigate('Delete Account')}>
              <Text style={styles.userSettingsActions}>
                {t('common.delete-account')}
              </Text>
            </TouchableOpacity>
          </>
        ) : null}
      </View>
      {!state.token ? (
        <TouchableOpacity onPress={() => navigation.navigate('Login')}>
          <Text style={styles.button}>{t('common.login')}</Text>
        </TouchableOpacity>
      ) : (
        <TouchableOpacity onPress={() => handleSignOut()}>
          <Text style={styles.button}>{t('common.logout')}</Text>
        </TouchableOpacity>
      )}
    </StyledView>
  );
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: Colors.primary,
    paddingVertical: 5,
    paddingHorizontal: 15,
    borderWidth: 0,
    borderRadius: 14,
    color: '#000',
    fontWeight: '500',
    fontSize: 20,
    overflow: 'hidden',
    textAlign: 'center',
    width: '40%',
    marginLeft: 10,
  },
  userSettingsContainer: {
    alignItems: 'flex-start',
  },
  userSettingsActions: {
    fontSize: 20,
    fontWeight: '500',
    lineHeight: 40,
  },
  userSettingsActionsView: {
    marginVertical: 50,
    marginLeft: 10,
  },
});
