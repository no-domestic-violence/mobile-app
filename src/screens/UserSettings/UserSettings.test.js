import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { AuthContext, LanguageProvider } from 'state';
import UserSettings from './index';

const mockLoggedOutState = {
    isLoggedIn: false,
    isFirstLaunch: false,
}
const mockLoggedInState = {
    isLoggedIn: true,
    errorMessage: '',
    isFirstLaunch: false,
    username: 'testuser',
    token: 'testjwttoken',
};
  
const mockNavigation = {
    navigate: jest.fn()
}
const signout = jest.fn();

describe('<UserSettings />', () => {    
    it('should match snapshot', () => {
        const result = render(
            <LanguageProvider>
                <AuthContext.Provider value={{ state: mockLoggedInState, signout }}>
                        <UserSettings />
                </AuthContext.Provider>
            </LanguageProvider>
        ).toJSON();
        expect(result).toMatchSnapshot();
    })
    it('should render user settings elements for logged in user', () => {
        const { getByText } = render(
                <AuthContext.Provider value={{ state: mockLoggedInState, signout }}>
                    <UserSettings />
                </AuthContext.Provider>

    );
        expect(getByText('common.logout')).toBeTruthy();
        expect(getByText('common.how-to-use')).toBeTruthy();
        expect(getByText('common.change-password')).toBeTruthy();
        expect(getByText('common.change-language')).toBeTruthy();
        expect(getByText('common.delete-account')).toBeTruthy();
    })

    it('should navigate to change password screen when change password button is pressed', () => {
        const { getByText } = render(
            <AuthContext.Provider value={{ state: mockLoggedInState, signout }}>
                <UserSettings navigation={mockNavigation} />
            </AuthContext.Provider>
        );
        fireEvent.press(getByText('common.change-password'));
        expect(mockNavigation.navigate).toHaveBeenCalledWith('Change Password');
    })

    it('should navigate to change language screen when change language button is pressed', () => {
        const { getByText } = render(
            <AuthContext.Provider value={{ state: mockLoggedInState, signout }}>
                <UserSettings navigation={mockNavigation} />
            </AuthContext.Provider>
        );
        fireEvent.press(getByText('common.change-language'));
        expect(mockNavigation.navigate).toHaveBeenCalledWith('Change Language');
    })

    it('should navigate to delete account screen when delete account button is pressed', () => {
        const { getByText } = render(
            <AuthContext.Provider value={{ state: mockLoggedInState, signout }}>
                <UserSettings navigation={mockNavigation} />
            </AuthContext.Provider>
        );
        fireEvent.press(getByText('common.delete-account'));
        expect(mockNavigation.navigate).toHaveBeenCalledWith('Delete Account');
    })

    it('should handle signout', () => {
        const { getByText } = render(
            <AuthContext.Provider value={{ state: mockLoggedInState, signout }}>
                <UserSettings />
            </AuthContext.Provider>

        );
        fireEvent.press(getByText('common.logout'));
        expect(signout).toHaveBeenCalledTimes(1);
    });
    it('should render user settings elements for logged out user', () => {
        const { getByText, queryByText } = render(
            <AuthContext.Provider value={{ state: mockLoggedOutState, signout }}>
                <UserSettings />
            </AuthContext.Provider>

        );
        expect(getByText('common.login')).toBeTruthy();
        expect(getByText('common.how-to-use')).toBeTruthy();
        expect(queryByText('common.change-password')).toBeNull();
        expect(queryByText('common.change-language')).toBeNull();
        expect(queryByText('common.delete-account')).toBeNull();
    });

    it('should navigate to logIn screen when log in button is pressed', () => {
        const { getByText } = render(
            <AuthContext.Provider value={{ state: mockLoggedOutState, signout }}>
                <UserSettings navigation={mockNavigation} />
            </AuthContext.Provider>
        );
        fireEvent.press(getByText('common.login'));
        expect(mockNavigation.navigate).toHaveBeenCalledWith('Login');
    })
});
