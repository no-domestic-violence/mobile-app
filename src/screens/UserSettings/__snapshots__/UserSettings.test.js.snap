// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserSettings /> should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#cadeee",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "alignItems": "flex-start",
      },
    ]
  }
>
  <View>
    <svg
      data-file-name="SvgUserSettings"
      style={
        Object {
          "position": "absolute",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginLeft": 10,
          "marginTop": 10,
        }
      }
    >
      <
        color="grey"
        icon={
          Object {
            "icon": Array [
              496,
              512,
              Array [],
              "f2bd",
              "M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z",
            ],
            "iconName": "user-circle",
            "prefix": "fas",
          }
        }
        size={80}
      />
      <Text
        style={
          Object {
            "fontSize": 28,
            "marginLeft": 15,
            "textTransform": "capitalize",
          }
        }
      >
        common.hello_user
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginLeft": 10,
        "marginVertical": 50,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "500",
            "lineHeight": 40,
          }
        }
      >
        common.how-to-use
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "500",
            "lineHeight": 40,
          }
        }
      >
        common.change-password
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "500",
            "lineHeight": 40,
          }
        }
      >
        common.change-language
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "500",
            "lineHeight": 40,
          }
        }
      >
        common.delete-account
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "#cadeee",
          "borderRadius": 14,
          "borderWidth": 0,
          "color": "#000",
          "fontSize": 20,
          "fontWeight": "500",
          "marginLeft": 10,
          "overflow": "hidden",
          "paddingHorizontal": 15,
          "paddingVertical": 5,
          "textAlign": "center",
          "width": "40%",
        }
      }
    >
      common.logout
    </Text>
  </View>
</View>
`;
